psycopg3/types/datetime.py

```diff
@@ -0,0 +1,112 @@
+"""
+Adapters for datetime types.
+https://www.postgresql.org/docs/9.4/datatype-datetime.html#DATATYPE-DATETIME-TABLE
+"""
+
+# Copyright (C) 2020 The Psycopg Team
+
+import codecs
+import datetime
+from typing import Union, Optional, Tuple
+
+from ..adapt import Dumper, Loader
+from ..proto import AdaptContext, DecodeFunc
+from .oids import builtins
+
+
+_encode = codecs.lookup("ascii").encode
+_decode = codecs.lookup("ascii").decode
+
+
+_time_oids = (
+    builtins["time"].oid,
+    builtins["timetz"].oid,
+    builtins["timestamp"].oid,
+    builtins["timestamptz"].oid,
+)
+
+
+_timezone_oids = (
+    builtins["timetz"].oid,
+    builtins["timestamptz"].oid,
+)
+
+
+@Dumper.text(datetime.date)
+@Dumper.binary(datetime.date)
+def dump_date(obj: datetime.date) -> Tuple[bytes, int]:
+    return _encode(str(obj))[0], builtins["date"].oid
+
+
+@Dumper.text(datetime.time)
+@Dumper.binary(datetime.time)
+def dump_time(obj: datetime.time) -> Tuple[bytes, int]:
+    if obj.tzinfo is None:
+        return _encode(str(obj))[0], builtins["time"].oid
+    else:
+        return _encode(str(obj))[0], builtins["timetz"].oid
+
+
+@Dumper.text(datetime.datetime)
+@Dumper.binary(datetime.datetime)
+def dump_datetime(obj: datetime.datetime) -> Tuple[bytes, int]:
+    if obj.tzinfo is None:
+        return _encode(str(obj))[0], builtins["timestamp"].oid
+    else:
+        return _encode(str(obj))[0], builtins["timestamptz"].oid
+
+
+@Dumper.text(datetime.timedelta)
+@Dumper.binary(datetime.timedelta)
+def dump_timedelta(obj: datetime.timedelta) -> Tuple[bytes, int]:
+    return _encode(str(obj))[0], builtins["interval"].oid
+
+
+@Loader.text(builtins["date"].oid)
+@Loader.binary(builtins["date"].oid)
+@Loader.text(builtins["time"].oid)
+@Loader.binary(builtins["time"].oid)
+@Loader.text(builtins["timetz"].oid)
+@Loader.binary(builtins["timetz"].oid)
+@Loader.text(builtins["timestamp"].oid)
+@Loader.binary(builtins["timestamp"].oid)
+@Loader.text(builtins["timestamptz"].oid)
+@Loader.binary(builtins["timestamptz"].oid)
+class DateTimeLoader(Loader):
+
+    decode: Optional[DecodeFunc]
+
+    def __init__(self, oid: int, context: AdaptContext):
+        super().__init__(oid, context)
+
+        if self.connection is not None:
+            if self.connection.encoding != "SQL_ASCII":
```

Comment:
We don't need special-casing sqlascii database. datetime formats are ascii only, unless I'm mistaken.

Reply:
I see. I wasn't quite sure what it was for in the text type since I'm not too well versed in jumping too far into PostgreSQL.

Reply:
For text, it's a special case for the `SQL_ASCII` format, because, rather than being "the database is encoded in ASCII", it means "the database has no encoding". That means it's pretty much a binary database: people can throw any byte between 1 and 255 at it and expect to find it back as it was. As such, for text, we don't decode the bytes received from database to an unicode `str`, like we normally do for `text` data type but we return them as python `bytes`.

This only happens for text types though: dates are only in formats including numbers and symbols < 128, so for them we don't really care about the encoding: we can expect that decoding from ASCII will work, and if it doesn't there is some other error somewhere.